<blockquote><h3>fabric 自动部署和运维工具</h3>

<p>fabric是python的ssh命令行工具, 可执行本地和远程的shell命令,
一般用来远程管理和持续化部署, 特别是在需要管理多台远程服务器的时候.</p>

<p>最简单的使用方法是用with cd(&ldquo;your dirpath&rdquo;)进入某个目录 然后执行run(&ldquo;some command&rdquo;)
code:
    with cd(&ldquo;/home/admin&rdquo;):
        run(&ldquo;./run.sh&rdquo;)</p>

<p>但是在使用过程中,有时候并无法正确的执行响应的任务:
在执行 ls, mkdir等命令或一些简单的脚本的时候可以正确执行任务,
但当启动某一个大型的工程的时候, 却没能正常启动,</p>

<p>在run.sh脚本中是这样启动的:
    nohup python manage.py run &amp;
在fab.py脚本中这样的运行的:
    run(&ldquo;./run.sh&rdquo;)</p>

<p>但是这样是不能启动的</p>

<p>这个问题困扰了我很久,网上有人说是:是由于Fabric过早关闭了连接的Session导致的
但是fabric在什么时候关闭的连接我没有搞清楚
于是我在run.sh脚本中执行网启动命令之后sleep一下
    nohup python manage.py run &amp;
    sleep 1</p>

<p>在执行run(&ldquo;./run,sh&rdquo;),问题解决</p>

<p>后来又找到一种方法,是这样执行:
run.sh:
    nohup python manage.py run &amp;</p></blockquote>

<p>fab.py:
    run(&ldquo;nohup ./run.sh&rdquo;)</p>

<blockquote><p>问题也能解决, 于是我认为是这样的:
fabric在执行完run中的命令之后,就关闭了连接
于是我就想把他放到后台去执行 用nohup去忽略所有挂断(SIGHUP)信号
或者 在执行启动命令之后sleep一段时间,给manage.py充分时间去启动,在退出脚本</p>

<p>但是启动还有一个疑问  我在启动manage.py的时候也有用到nohup,但是为什么不起作用,这点我不是很清楚</p></blockquote>
